api/routes.go

```diff
@@ -85,15 +92,29 @@ func (r *RouteHandler) recommendClusterSetup(c *gin.Context) {
 	log.Info("recommend cluster setup")
 	provider := c.Param("provider")
 	region := c.Param("region")
-	var request recommender.ClusterRecommendationReq
-	if err := c.BindJSON(&request); err != nil {
-		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+
+	// request decorated with provider and region
+	reqWr := RequestWrapper{P: provider, R: region}
+
+	if err := c.BindJSON(&reqWr); err != nil {
+		c.JSON(http.StatusBadRequest, map[string]string{
+			"code":    "bad_params",
+			"message": "invalid zone",
+			"cause":   err.Error(),
+		})
 		return
 	}
-	// TODO: validation
-	if response, err := r.engine.RecommendCluster(provider, region, request); err != nil {
+
+	if response, err := r.engine.RecommendCluster(provider, region, reqWr.ClusterRecommendationReq); err != nil {
 		c.JSON(http.StatusInternalServerError, gin.H{"status": http.StatusInternalServerError, "message": fmt.Sprintf("%s", err)})
 	} else {
 		c.JSON(http.StatusOK, *response)
 	}
 }
+
+// RequestWrapper internal struct for passing provider/zone info to the validator
+type RequestWrapper struct {
```

Comment:
I would change it to lowercase if it's an internal struct if package visibility makes it possible.

Reply:
thx, fixed